llvm/unittests/CodeGen/SelectionDAGPatternMatchTest.cpp

```diff
@@ -119,6 +119,26 @@ TEST_F(SelectionDAGPatternMatchTest, matchValueType) {
   EXPECT_FALSE(sd_match(Op2, m_ScalableVectorVT()));
 }
 
+TEST_F(SelectionDAGPatternMatchTest, matchVecShuffle) {
+  SDLoc DL;
+  auto Int32VT = EVT::getIntegerVT(Context, 32);
+  auto VInt32VT = EVT::getVectorVT(Context, Int32VT, 4);
+  SmallVector<int, 4> MaskData = {2, 0, 3, 1};
+  ArrayRef<int> Mask;
+
+  SDValue V0 = DAG->getCopyFromReg(DAG->getEntryNode(), DL, 1, VInt32VT);
+  SDValue V1 = DAG->getCopyFromReg(DAG->getEntryNode(), DL, 2, VInt32VT);
+  SDValue VecShuffleWithMask_0 =
+      DAG->getVectorShuffle(VInt32VT, DL, V0, V1, MaskData);
+
+  using namespace SDPatternMatch;
+  EXPECT_TRUE(
+      sd_match(VecShuffleWithMask_0, m_Shuffle(m_Value(V0), m_Value(V1))));
+  EXPECT_TRUE(sd_match(VecShuffleWithMask_0,
+                       m_Shuffle(m_Value(V0), m_Value(V1), Mask)));
+  EXPECT_TRUE(std::equal(Mask.begin(), Mask.end(), MaskData.begin()));
```

Comment:
please see my comment above, I think the test here is wrong.

Reply:
You might wonder why the pre-merge CI passed, and I think that's because `std::equal` is likely to be implemented like this:
```
template<class InputIt1, class InputIt2>
constexpr //< since C++20
bool equal(InputIt1 first1, InputIt1 last1, InputIt2 first2)
{
    for (; first1 != last1; ++first1, ++first2)
        if (!(*first1 == *first2))
            return false;
 
    return true;
}
```
In your case, since `Mask` is empty, `Mask.begin()` is equal to `Mask.end()`, hence the true return value.

Try `std::equal(MaskData.begin(), MaskData.end(), Mask)` or `std::equal(Mask.begin(), Mask.end(), MaskData.begin(), MaskData.end())` and this test probably won't pass.

Reply:
I resorted, in the refactored version, to simply doing `I->getMask() == SpecificMask`.  As `m_Mask` is no longer used, I believed the test for the equality of `Mask` and `MaskContents` after the call to `m_Mask` was no longer needed. The test for contents only matters in the `m_ShuffleSpecificMask` case?